src/GovUk.Education.ExploreEducationStatistics.Data.Api/Services/PermalinkService.cs

```diff
@@ -97,14 +109,38 @@ private static JsonSerializerSettings BuildJsonSerializerSettings()
         private async Task<PermalinkViewModel> BuildViewModel(Permalink permalink)
         {
             var subject = await _subjectRepository.Get(permalink.Query.SubjectId);
-            var isValid = subject != null && await _subjectRepository.IsSubjectForLatestPublishedRelease(subject.Id);
+
+            var isValid = subject != null
+                          && !IsSuperseded(subject.Id)
+                          && await _subjectRepository.IsSubjectForLatestPublishedRelease(subject.Id);
 
             var viewModel = _mapper.Map<PermalinkViewModel>(permalink);
 
             viewModel.Invalidated = !isValid;
 
             return viewModel;
         }
+
+        private bool IsSuperseded(Guid subjectId)
+        {
+            var releaseId = _statisticsDbContext.ReleaseSubject
+                .Where(rs => rs.SubjectId == subjectId)
+                .Select(rs => rs.ReleaseId)
+                .Single();
+
+            var publication = _contentDbContext.Releases
+                .Include(r => r.Publication)
+                .Where(r => r.Id == releaseId)
+                .Select(r => r.Publication)
+                .Single();
+
+            return publication.SupersededById != null
+                   && _contentDbContext.Releases
+                       .Include(p => p.Publication)
+                       .Where(r => r.PublicationId == publication.SupersededById)
+                       .ToList()
+                       .Any(r => r.IsLatestPublishedVersionOfRelease());
+        }
```

Comment:
Seems like we're doing a lot of the same queries that are being performed in `_subjectRepository.IsSubjectForLatestPublishedRelease`. This is pretty inefficient as we're doing a lot of unnecessary database calls.

Is it possible to combine these two? Or at the very least, consolidate this logic into `_subjectRepository` with a new method?

Reply:
The current code does make a DB call to get the `ReleaseSubject` twice
, so it's possible I could make that call once and then pass the `ReleaseSubject` to the relevant methods - but the downside to making one less call is the code will be harder to understand.




I did consider moving `PermalinkService#IsSuperseded` to `SubjectRepository` - but that would have required adding `ContentDbContext` to `SubjectRepository` and so decided against it

, but maybe that is fine?


On an aside, I'm not actually a fan of `SubjectRepository#IsSubjectForLatestPublishedRelease`, as it only checks the Stats DB, and that means that if a new release is published without any subject, then no new rows get created in the Stats DB, and the method actually says the subject is for the latest published release when it isn't

. I think we've accepted this so far because it is expected that all releases will have subjects attached - but it seems plausible that in the future that may not be the case - I'd prefer to cover that eventuality now.

I'm inclined to just leave this. This code is going to get changed anyway with the permalink error stuff (EES-3143). But if you still want me to , I'm happy to pull the `ReleaseSubject` DB call out so it only gets made once and/or move the `IsSuperseded` method to `SubjectRepository`.

Reply:
Ben suggested how I should go about this on Slack so I've done that - probably just over complicating it in my mind.